### ISSUE TITLE
Speed up `z_sendmany`, which currently takes ~3 seconds to complete.

### ISSUE THREAD

**Issue by holmesworcester** [zcash/zcash#4429]
## Is your feature request related to a problem? Please describe.
Our app Zbay uses the encrypted memo field to send messages in a chat-like interface, so the speed of sending a message is very important to the UX. We are using the `z_sendmany` RPC call to send funds to a zaddr with memo field included, and after adding some timestamps in the zcashd source code, we noticed that `z_sendmany` takes around three seconds to complete. 

## Describe the solution you'd like
A speed improvement of 3x or more, so that `z_sendmany` takes 1 second or less to complete.

## Alternatives you've considered
We've taken other steps to improve the speed of delivery of messages, but this call is now the biggest factor, so other optimizations outside this call will have insignificant impact.

## Additional context

Here are some examples of sending a message, with timestamps. 

```
1584714239611:
opid-1c3a015a-0084-4970-8bc1-6406fcfbacf9: z_sendmany initialized (params={"fromaddress":"zs1sc4xrr5r00cv8fh83k528ec4ua9p6mu834mg46jgx8qf9metcweg53gmw7fdrt8ygq3c747qztt","amounts":[{"address":"zs1xp8psckj029fuex7glfmz0yv7sk8gwk7e620d9auxfnq2kukypwnrt8v84cayvhvygaku7lswfz","amount":"0","memo":"ff01d31a1e7ac23eb9751de7eca95e40b07a59420d4f07f26bb12184dded7a9e9f5c1bffd1345ad53f5e320434965d1b83fbee767ea2898a1a0e73e528f9758b43ad015e74d1ff654a78547973685541674143754145570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"}],"minconf":1,"fee":2.5e-06})
1584714243328:
_opid-1c3a015a-0084-4970-8bc1-6406fcfbacf9: z_sendmany finished (status=success, txid=24bd89fc437c520fe2ab62c63f3b80a95d06e5daeacededdf38c0d45c449b911)
```

**Comment by daira** [comment]
The cost of creating a shielded transaction is primarily dependent on the number of shielded inputs (shielded outputs are much less expensive). Although there are potential optimizations that we could do for proving in bellman, they would be very unlikely to provide a factor-of-3 improvement; improvements in the 10s of percent would be more likely, unless and until we do another major upgrade like Sapling that moves to a different proving system.

**Comment by holmesworcester** [comment]
Thanks for your reply! Improvements in the 10s of percent would not be meaningful to us, so this is good to know.

> The cost of creating a shielded transaction is primarily dependent on the number of shielded inputs

If a shielded transaction is created directly from a transparent input, will it be faster?

**Comment by str4d** [comment]
Sapling spends are around 10x slower to create than Sapling outputs. For the example transaction you referenced in the OP, it has 1 Sapling Spend description and two Sapling Output descriptions, so it would have been something like 2.5s to make the Spend description, and 0.25s for each of the Output descriptions.

So yes, spending transparent funds would be faster. However, you would then be creating T->Z transactions, not Z->Z transactions. You would need to think very carefully about the information leakage this causes (for example, all message transactions sent by a user become trivially linkable, so e.g. the volume of messaging conducted by a user is clearly observable), and whether that is a privacy trade-off you can safely convey to your users.

IMHO, a three-second delay is small enough that it can be covered up by the UI. I expect most users would be fine with a three-second latency in a text conversation (replying to a message could easily take longer than that). But if they weren't, then an alternative would be that the UI detects when a user starts writing a message, and pre-emptively prepares the Sapling spends. This isn't something that can be done through the current `zcashd` RPC API, but there's no technical barrier to implementing this. If provided for in `zcashd`, it would be something like:
- A transaction is "started" by making a call to `zcashd`, passing in the maximum amount that will be spent. Sufficient notes are selected, and the Spend proofs begin being computed. Some handle is returned to the caller representing the prepared transaction.
- A prepared transaction is "finished" by making a separate call to `zcashd`, passing in the handle, and the full list of recipients, values, and memos. This call would have several possible results:
  - If there was too long a gap between the two calls (e.g. if enough blocks have gone by that the transaction's Sapling anchor needs to be redone, or some of the selected notes have been detected as spent), the call fails.
  - If the provided recipients require more funds than was prepared, then the call could fail, or could maybe optionally continue on (by selecting additional funds, at the cost of needing to make Spend descriptions for them).
  - Otherwise, the transaction sending process continues in the same async way as `z_sendmany`, returning an opid that can be checked for completion later.

**Comment by holmesworcester** [comment]
Yes, this approach would definitely work for us. 

As far as UX goes, 3 seconds is still a little too much if you're talking to someone, and on a less powerful device it'd be potentially a lot more than 3 seconds, so if we can address it we would like to.

What would be the best way to move forward? Could we write up some proposed improvements to the RPC API that would let us do what we want, get feedback, and pursue them?

**Comment by str4d** [comment]
Yep, that sounds good to me. It might also be useful to write an informational ZIP that describes the motivation and specifies the workflow. In either case, we definitely want to get feedback from other potential users, so I'd also recommend that once there's a reasonable design for the API, it be implemented and merged as an experimental feature first (which makes it easy for everyone interested to test the functionality, while giving us the flexibility to make breaking changes to it at any time as we improve it).

**Comment by holmesworcester** [comment]
Great. We'll discuss this internally and proceed from there!

**Comment by kowalski** [comment]
@jkdev601 spend today some time profiling the operation. We've identified a time consuming step in here:

src/wallet/asyncrpcoperation_sendmany.cpp:935
```
bool AsyncRPCOperation_sendmany::find_unspent_notes() {
    std::vector<SproutNoteEntry> sproutEntries;
    std::vector<SaplingNoteEntry> saplingEntries;
    {
        LOCK2(cs_main, pwalletMain->cs_wallet);
        pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, fromaddress_, mindepth_);
    }
    // If using the TransactionBuilder, we only want Sapling notes.
    // If not using it, we only want Sprout notes.
    // TODO: Refactor `GetFilteredNotes()` so we only fetch what we need.
```

Janusz put the log lines around invocation of `GetFilteredNotes`:

Here are the results:

```
80609_STARTS_FINDIND_UNSPENT_NOTESReceived a POST request for / from 127.0.0.1:32822 
81156_FINISHED_FINDING_UNSPENT_NOTESopid-b3f25ad0-e270-4c2f-b4aa-e9a6a929c92c: found unspent Sapling note (txid=0b1a35471f, vShieldedSpend=0, amount=0.00, memo=ff0197e05d)
81157_STARTS_SORTING_NOTES81157_FINISHED_SORTING_NOTESopid-b3f25ad0-e270-4c2f-b4aa-e9a6a929c92c: spending 0.0000025 to send 0.00 with fee 0.0000025
83404_Z_SENDMANY_FINISHED_opid-b3f25ad0-e270-4c2f-b4aa-e9a6a929c92c: z_sendmany finished (status=success, txid=008fa33505f86264317645a0c22c4c43b81573cca7454319a408071ef73d2109)
```

Number in front is number of miliseconds since starting the process. So the call to `GetFilteredNotes` took 547ms out of 2795ms of the complete call, so it's responsible for almost 20% of the entire time `z_sendmany` took.

I think the reason it takes so long, is that it iterates over all the notes known to the node and for each note it calls:
```
            auto maybe_pt = SaplingNotePlaintext::decrypt(
                wtx.vShieldedOutput[op.n].encCiphertext,
                nd.ivk,
                wtx.vShieldedOutput[op.n].ephemeralKey,
                wtx.vShieldedOutput[op.n].cmu);
```
to than extract payment address and compare it with `fromaddress_`. 
It's done every single time. 

In Zbay we have a lot of notes, because for each message sent we receive a note "with a change" and also each incoming message is also a note. So our local node has to deal with more notes than most notes. 
To make things worse, the list of notes grows over time without bounds, leading to deteriorating performance over time.


I think a possible solution could be to add in-memory index btree structure that would allow us to make range queries very efficiently. This is directly related with  https://github.com/zcash/zcash/issues/4394 in which we also suffer from performance issues caused by implementation on `GetFilteredNotes()`. In that issue we've explained that we would like to make queries for notes belonging to an address within a timestamp range. I suggested creating an index structure using fields `[paymentAddress, timestamp]`. The same structure would do a great job for use in `AsyncRPCOperation_sendmany::find_unspent_notes() `. In this case we would query it using range from `[fromaddress_, 0]` to `[from_address_, INT_MAX]`. 

Please let me know what you think.

**Comment by oxarbitrage** [comment]
Thank you for the report @kowalski , the same performance issue with `GetFilteredNotes()` is also present in other issues as well so we should discuss here(or in a dedicated issue) how is the best way to handle it. 

Implementation idea:
I think we can index the fields we need with every transaction coming into the wallet in `AddToWallet()`. If needed, we can do the decryption there and store in a new object or ideally just index on existing `CWalletTx`. I am a bit concerned on the security of doing this.

I am exploring the boost::multi_index option in the context of https://github.com/zcash/zcash/issues/4394

**Comment by kowalski** [comment]
> I am a bit concerned on the security of doing this.

Perhaps the index should be on on `[hash(payment_address), timestamp]`, instead of just address, 
where `hash(x)` is some one-way function.
We could do it both from security concerns and limiting memory usage of this structure. We could for example only take 4 bytes of the address hash and 4 bytes for timestamp (8 bytes in total per Note). We can handle collisions by filtering in the limited dataset.

**Comment by str4d** [comment]
> I think the reason it takes so long, is that it iterates over all the notes known to the node and for each note it calls `SaplingNotePlaintext::decrypt`

It should only be calling that for spendable notes, not all known notes. Having said that, I wonder whether zero-value notes are being included; while you can technically spend a zero-value note, no one would ever do so, and `GetFilteredNotes()` should only be returning notes that can contribute value to a transaction. I presume that you are mostly using zero-value notes for your application?

**Comment by kowalski** [comment]
>  I presume that you are mostly using zero-value notes for your application?

I don't think these are zero-value notes, beacuse they are enough to cover transaction fees. But I'm not sure. @jkdev601 @NorbertBodziony can one of you answer this part?


When I look in the logs I can see it's reporting logs like:
```
2020-04-08 11:38:20 opid-b3f25ad0-e270-4c2f-b4aa-e9a6a929c92c: found unspent Sapling note (txid=74c1dabdba, vShieldedSpend=0, amount=0.00, memo=ff0146bd4e)
```
I counted 37 lines like this in this invocation, so we do quite a few of them.

**Comment by NorbertBodziony** [comment]
> I presume that you are mostly using zero-value notes for your application?

Yes all messages we send include `amount = 0` so we are generating zero-value notes.

**Comment by str4d** [comment]
I double-checked, and I was mis-remembering; we do indeed trial-decrypt every Sapling note. In general, we currently only cache per-note data that is necessary, and decrypt to access anything we already inherently "store" in the transaction itself.
- For Sprout notes we cache the payment address, because we need it to be able to decrypt the note. We can therefore use it for filtering.
- For Sapling notes, we only cache `ivk` because that is all we need for decryption, but we then need to decrypt in order to access the payment address for filtering.
- We don't cache the value in either case.

For this particular efficienty issue, we would benefit from caching the payment address (which could happen indirectly as part of the lookup structure being discussed above), and whether or not the note is zero-value (which could be done by adding a bitfield to `SaplingNoteData` that is populated during trial-decryption).

**Comment by kowalski** [comment]
@str4d that sounds right to me.
